lib/ain-evm/src/core.rs

```diff
@@ -740,6 +755,88 @@ impl EVMCoreService {
         trace!("[get_next_valid_nonce_in_block_template] Account {address:x?} nonce {nonce:x?}");
         Ok(nonce)
     }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn trace_transaction(
+        &self,
+        caller: H160,
+        to: H160,
+        value: U256,
+        data: &[u8],
+        gas_limit: u64,
+        access_list: AccessList,
+        block_number: U256,
+    ) -> Result<(Vec<ExecutionStep>, bool, Vec<u8>, u64)> {
+        let (state_root, block_number) = self
+            .storage
+            .get_block_by_number(&block_number)?
+            .ok_or_else(|| format_err!("Cannot find block"))
+            .map(|block| (block.header.state_root, block.header.number))?;
+
+        debug!(
+            "Calling EVM at block number : {:#x}, state_root : {:#x}",
+            block_number, state_root
+        );
+
+        let vicinity = Vicinity {
+            block_number,
+            origin: caller,
+            ..Default::default()
+        };
```

Comment:
Not setting the full vicinity state of the backend with the block context could lead to incorrect state execution

Reply:
Resolved on PR https://github.com/DeFiCh/ain/pull/2641